### PR TITLE
WalletConnect Provider v4.0.0

### PR DESCRIPTION
**Pull request by radumojic**
- Updated @WalletConnect to `2.8.6` 
- Added the nativeAuthToken connect functionality that will send a different request based on the login token type
    - `mvx_signNativeAuthToken` for native auth compatible tokens
    - `mvx_signLoginToken` for any other token types ( will be deprecated in the future )
- Prepare for Guardians functionality: accept `guardian`, `guardianSignature`, `options` and `version` besides the `signature` in the response from the device

**Breaking Changes:**

- Drop interfaces, add dependency on [sdk-core](https://www.npmjs.com/package/@multiversx/sdk-core)
- Dropped WalletConnect's v1 SDKs as they are now deprecated and it will be sunset soon: https://medium.com/walletconnect/walletconnect-v1-0-sunset-notice-and-migration-schedule-8af9d3720d2e
- Drop `namespace`, `events` and `methods` from the provider class as they are no longer needed in the functionality

Although these changes are intrinsically breaking, the transition should be pretty seamless since any changes in the basic workflow would remain backwards compatible and the new functionalities would only enhance the present functionality

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  const acknowledgedSessions = client
+    .find(getConnectionParams(chainId))
+    .filter((s) => s.acknowledged);
+
+  if (acknowledgedSessions.length > 0) {
+    const lastKeyIndex = acknowledgedSessions.length - 1;
+    const session = acknowledgedSessions[lastKeyIndex];
+
+    return session;
+  } else if (client.session.length > 0) {
```
Comment (andreibancioiu) [inline src/utils.ts:36]: `else` statements can be removed (can also stay as they are).

File: src/walletConnectV2Provider.ts [hunk 2]
```diff
@@ -1,21 +1,26 @@
+import { IPlainTransactionObject } from "@multiversx/sdk-core/out";
+import { Transaction } from "@multiversx/sdk-core/out/transaction";
```
Comment (andreibancioiu) [inline src/walletConnectV2Provider.ts:2]: Can we import without the `/out/` segment?

File: src/walletConnectV2Provider.ts [hunk 3]
```diff
@@ -200,13 +195,22 @@ export class WalletConnectV2Provider {
...
+          const address = getAddressFromSession(session);
+
+          const selectedNamespace =
+            session.namespaces[WALLETCONNECT_MULTIVERSX_NAMESPACE];
+          const method = selectedNamespace.methods.includes(
+            OptionalOperation.SIGN_NATIVE_AUTH_TOKEN
+          )
+            ? OptionalOperation.SIGN_NATIVE_AUTH_TOKEN
+            : OptionalOperation.SIGN_LOGIN_TOKEN;
+
```
Comment (andreibancioiu) [inline src/walletConnectV2Provider.ts:207]: If the obsolete login is being used, can we display a WARN message?

File: src/walletConnectV2Provider.ts [hunk 4]
```diff
@@ -422,32 +428,36 @@ export class WalletConnectV2Provider {
...
       }
 
-      transaction.applySignature(
-        Signature.fromHex(signature),
-        UserAddress.fromBech32(address)
-      );
-      return transaction;
+      const responseKeys = Object.keys(response);
+      // if on old flow and only signature is present
+      if (responseKeys.length === 1 && responseKeys[0] === "signature") {
```
Comment (andreibancioiu) [inline src/walletConnectV2Provider.ts:454]: Add a temporary variable, such as `responseIsWholeTx` / `responseIsJustThePlainOldSignature` (these are merely examples)?

File: src/walletConnectV2Provider.ts [hunk 5]
```diff
@@ -488,42 +494,71 @@ export class WalletConnectV2Provider {
...
         );
       }
 
-      for (const [index, transaction] of transactions.entries()) {
-        transaction.applySignature(
-          Signature.fromHex(signatures[index].signature),
-          UserAddress.fromBech32(address)
+      const responseKeys = Object.keys(response);
+      // if on old flow and only signatures is present
+      if (responseKeys.length === 1 && responseKeys[0] === "signatures") {
```
Comment (andreibancioiu) [inline src/walletConnectV2Provider.ts:524]: As above, a named temporary variable might be useful (for code reading).

File: src/walletConnectV2Provider.ts [hunk 6]
```diff
@@ -488,42 +494,71 @@ export class WalletConnectV2Provider {
...
+        const { signatures } = response as TransactionsResponseOldType;
+        if (
```
Comment (andreibancioiu) [inline src/walletConnectV2Provider.ts:526]: Can we extract the logic to separate functions (old flow vs. new flow)? Leave as it is, though, if it adds more complexity.

File: src/utils.ts (hunk 1 above)
Comment (radumojic) [reply src/utils.ts]: Updated

File: src/walletConnectV2Provider.ts (hunk 6 above)
Comment (radumojic) [reply src/walletConnectV2Provider.ts]: Changed the transaction workflow, no longer needed.

File: src/walletConnectV2Provider.ts (hunk 5 above)
Comment (radumojic) [reply src/walletConnectV2Provider.ts]: Changed the transaction workflow, no longer needed.

File: src/walletConnectV2Provider.ts (hunk 3 above)
Comment (radumojic) [reply src/walletConnectV2Provider.ts:203]: I think it's best that we add this warning once we have the nativeAuth login flow documented

File: src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  if (transactionGuardian && transactionGuardian !== guardian) {
+    Logger.error(WalletConnectV2ProviderErrorMessagesEnum.invalidGuardian);
+    throw new Error(WalletConnectV2ProviderErrorMessagesEnum.invalidGuardian);
+  }
+
+  if (guardian) {
+    transaction.setGuardian(Address.fromBech32(guardian));
+  }
+
+  if (version !== undefined) {
```
Comment (CiprianDraghici) [inline src/utils.ts:155]: `version` could be null ? If yes, then a better condition could be `version != null`

File: src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    transaction.setVersion(version);
+  }
+
+  if (options !== undefined) {
```
Comment (CiprianDraghici) [inline src/utils.ts:159]: Same here...see the previous [comment](https://github.com/multiversx/mx-sdk-js-wallet-connect-provider/pull/26/files#r1251915974)

File: src/utils.ts (hunk 7 above)
Comment (radumojic) [reply src/utils.ts]: updated to `if(version)` since version can only be >= 1

File: src/utils.ts (hunk 8 above)
Comment (radumojic) [reply src/utils.ts]: updated to `version != null`

**Review by CiprianDraghici — DISMISSED** [review]

**Review by andreibancioiu — APPROVED** [review]
Some functions could have been unit tested.

File: src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+  Logger.error(WalletConnectV2ProviderErrorMessagesEnum.sessionNotConnected);
+  throw new Error(WalletConnectV2ProviderErrorMessagesEnum.sessionNotConnected);
+}
+
+export function getCurrentTopic(
+  client: Client,
+  chainId: string
+): SessionTypes.Struct["topic"] {
+  if (typeof client === "undefined") {
```
[the remark this answers is not in the document]
Comment (radumojic) [reply src/utils.ts]: updated

File: src/utils.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  const session = getCurrentSession(client, chainId);
+  if (session?.topic) {
+    return session.topic;
+  } else {
+    throw new Error(
+      WalletConnectV2ProviderErrorMessagesEnum.sessionNotConnected
+    );
+  }
```
[the remark this answers is not in the document]
Comment (radumojic) [reply src/utils.ts]: updated

File: src/utils.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        chains,
+        events,
+      },
+    },
+  };
+}
+
+export function addressIsValid(destinationAddress: string): boolean {
+  try {
+    const addr = Address.fromBech32(destinationAddress);
```
[the remark this answers is not in the document]
Comment (radumojic) [reply src/utils.ts]: updated with `!address.isEmpty()`

**Comment by radumojic** [comment]
> Some functions could have been unit tested.

Agree, will add some unit tests in the future based on the ones from @WalletConnect

**Review by andreibancioiu — APPROVED** [review]

**Review by CiprianDraghici — APPROVED** [review]

**Pull request merged**